src/Stratis.Bitcoin.Features.Consensus/PosConsensusValidator.cs

```diff
@@ -467,23 +425,112 @@ public void CheckAndComputeStake(ContextInformation context)
+                this.logger.LogTrace("(-)[P2PK]:{0}", res);
+                return res;
+            }
+
+            // Block signing key also can be encoded in the nonspendable output.
+            // This allows to not pollute UTXO set with useless outputs e.g. in case of multisig staking.
+
+            List<Op> ops = txout.ScriptPubKey.ToOps().ToList();
+            if (!ops.Any()) // script.GetOp(pc, opcode, vchPushValue))
+            {
+                this.logger.LogTrace("(-)[NO_OPS]:false");
+                return false;
+            }
+
+            if (ops.ElementAt(0).Code != OpcodeType.OP_RETURN) // OP_RETURN)
+            {
+                this.logger.LogTrace("(-)[NO_OP_RETURN]:false");
+                return false;
+            }
+
+            if (ops.Count < 2) // script.GetOp(pc, opcode, vchPushValue)
+            {
+                this.logger.LogTrace("(-)[NO_SECOND_OP]:false");
+                return false;
+            }
+
+            byte[] data = ops.ElementAt(1).PushData;
+            if (!ScriptEvaluationContext.IsCompressedOrUncompressedPubKey(data))
+            {
+                this.logger.LogTrace("(-)[NO_PUSH_DATA]:false");
+                return false;
+            }
+
+            bool verifyRes = new PubKey(data).Verify(block.GetHash(), new ECDSASignature(block.BlockSignatur.Signature));
+            this.logger.LogTrace("(-):{0}", verifyRes);
+            return verifyRes;
         }
 
-        // Miner's coin stake reward.
-        public Money GetProofOfStakeReward(int height)
+        /// <summary>
+        /// Checks whether the coinstake timestamp meets protocol.
+        /// </summary>
+        /// <param name="blockTime">The block time.</param>
+        /// <param name="transactionTime">The transaction time.</param>
+        /// <returns><c>true</c> if block timestamp is equal to transaction timestamp, <c>false</c> otherwise.</returns>
+        private bool checkCoinStakeTimestamp(long blockTime, long transactionTime)
         {
-            if (this.IsPremine(height))
-                return this.consensusOptions.PremineReward;
+            return (blockTime == transactionTime) && ((transactionTime & StakeTimestampMask) == 0);
+        }
 
-            return this.consensusOptions.ProofOfStakeReward;
+        /// <summary>
+        /// Checks whether the future drift should be reduced after provided timestamp.
+        /// </summary>
+        /// <param name="time">Timestamp.</param>
+        /// <returns><c>true</c> if for this timestamp future drift should be reduced, <c>false</c> otherwise.</returns>
+        private bool IsDriftReduced(long time)
+        {
+            return time > DriftingBugFixTimestamp;
+        }
+
+        /// <summary>
+        /// Applies future drift to provided timestamp.
+        /// </summary>
+        /// <remarks>
+        /// Future drift is maximal allowed block's timestamp difference over local time.
```

Comment:
over adjusted time, not local time

Reply:
At first I had something like "over local time (usually adjusted)" but changed it to that because by local we could mean that it's adjusted already. 

Just checked IDateTimeProvider so yeah, 'local' is not applicable here.